File: core/backend/app/sessions/condenser.py
from __future__ import annotations
import uuid
"""
SessionCondenser — pattern-based MVP for session → derived summary.

Invariants:
- Never creates a MemoryEntry (active or proposed).
- Never creates a Proposal.
- SessionSummary is derived context, not source of truth.
- Multiple summaries per session: latest active is the one used by ContextBuilder.
"""

import logging
import re
from collections import Counter
from datetime import UTC, datetime

from sqlalchemy.orm import Session as DBSession

from ..models import Message, SessionSummary

log = logging.getLogger(__name__)

_CONDENSER_VERSION = "pattern.v1"
_STOPWORDS = frozenset(
    "a an the is are was were be been being have has had do does did will would could should "
    "may might shall can i you he she it we they me him her us them my your his its our their "
    "this that these those and or but not in on at to of for with by from up out as so if then "
    "no yes just also like more some any all here there when where how what who".split()
)
_MAX_MESSAGES = 40
_MAX_SUMMARY_CHARS = 1_200


def _new_id() -> str:
    return str(uuid.uuid4())


def _extract_keywords(text: str, top_n: int = 8) -> list[str]:
    words = re.findall(r"[a-zA-Z]{3,}", text.lower())
    filtered = [w for w in words if w not in _STOPWORDS]
    return [w for w, _ in Counter(filtered).most_common(top_n)]


def _summarise_messages(messages: list[Message]) -> str:
    """
    Pattern-based condensation: extracts role counts, key user phrases, top keywords.
    Produces a 1–3 sentence plaintext summary. No LLM calls.
    """
    user_turns = [m for m in messages if m.role == "user"]
    assistant_turns = [m for m in messages if m.role == "assistant"]

    total = len(messages)
    lines: list[str] = [f"Session with {total} messages ({len(user_turns)} user, {len(assistant_turns)} assistant)."]

    # Extract first non-trivial user message as the stated goal
    goal_sentence = ""
    for m in user_turns[:3]:
        text = (m.content or "").strip()
        if len(text) > 20:
            # Trim to first sentence
            sentence = re.split(r"[.!?\n]", text)[0].strip()
            if sentence:
                goal_sentence = sentence[:200]
                break
    if goal_sentence:
        lines.append(f"User goal: {goal_sentence}.")

    # Keyword extraction across all content
    all_text = " ".join(
        (m.content or "")
        for m in messages
        if m.role in ("user", "assistant")
    )
    keywords = _extract_keywords(all_text)
    if keywords:
        lines.append(f"Key topics: {', '.join(keywords)}.")

    summary = " ".join(lines)
    return summary[:_MAX_SUMMARY_CHARS]


class SessionCondenser:
    """
    Condenses a session's messages into a derived SessionSummary.

    Never writes MemoryEntry, Proposal, or Policy.
    """

    def __init__(self, db: DBSession) -> None:
        self._db = db

    def get_latest(self, session_id: str, space_id: str) -> SessionSummary | None:
        """Return the current active summary for a session, or None."""
        return (
            self._db.query(SessionSummary)
            .filter(
                SessionSummary.session_id == session_id,
                SessionSummary.space_id == space_id,
                SessionSummary.status == "active",
            )
            .order_by(SessionSummary.version.desc())
            .first()
        )

    def condense(self, session_id: str, space_id: str, user_id: str | None = None) -> SessionSummary:
        """
        Condense the session's messages into a new SessionSummary.

        Supersedes any existing active summary for this session.
        Never creates MemoryEntry or Proposal.
        Queries messages filtered by both session_id and space_id.
        """
        messages = (
            self._db.query(Message)
            .filter(Message.session_id == session_id, Message.space_id == space_id)
            .order_by(Message.created_at.asc())
            .limit(_MAX_MESSAGES)
            .all()
        )

        if not messages:
            summary_text = "Empty session — no messages to summarise."
            first_msg_id = None
            last_msg_id = None
        else:
            summary_text = _summarise_messages(messages)
            first_msg_id = messages[0].id
            last_msg_id = messages[-1].id

        # Char-level token estimates (≈ 4 chars/token approximation)
        source_chars = sum(len(m.content or "") for m in messages)
        summary_chars = len(summary_text)

        keywords = _extract_keywords(
            " ".join(m.content or "" for m in messages if m.role in ("user", "assistant"))
        )
        user_turns = [m for m in messages if m.role == "user"]
        assistant_turns = [m for m in messages if m.role == "assistant"]
        summary_json_value: dict = {
            "condenser_version": _CONDENSER_VERSION,
            "role_counts": {"user": len(user_turns), "assistant": len(assistant_turns)},
            "top_keywords": keywords,
            "source_range": {
                "first_message_id": first_msg_id,
                "last_message_id": last_msg_id,
                "message_count": len(messages),
            },
        }

        # Supersede existing active summary
        existing = self.get_latest(session_id, space_id)
        new_version = 1
        if existing is not None:
            existing.status = "superseded"
            self._db.add(existing)
            new_version = existing.version + 1

        row = SessionSummary(
            id=_new_id(),
            space_id=space_id,
            session_id=session_id,
            user_id=user_id,
            version=new_version,
            status="active",
            summary_text=summary_text,
            source_message_count=len(messages),
            source_first_message_id=first_msg_id,
            source_last_message_id=last_msg_id,
            summary_json=summary_json_value,
            token_estimate_before=source_chars // 4,
            token_estimate_after=summary_chars // 4,
            condenser_version=_CONDENSER_VERSION,
            created_at=datetime.now(UTC),
        )
        self._db.add(row)
        self._db.flush()
        log.debug(
            "SessionCondenser: condensed session=%s v%d (%d msgs)",
            session_id,
            new_version,
            len(messages),
        )
        return row

File: server/src/modules/sessions/condenser.ts
/**
 * Session condensers.
 *
 * Two condenser versions produce a `SessionSummary` body from a slice of session
 * messages:
 *
 * - `pattern.v1` — deterministic, no LLM. A pure function (`buildPatternSummary`)
 *   of role counts, top keywords, and a compact highlight transcript. Free,
 *   instant, fully reproducible, but it does not understand meaning. Used as the
 *   always-available fallback.
 * - `llm.v1` — an LLM writes a real running summary from a prompt resolved
 *   through the centralized prompt registry; `buildLlmSummary` wraps the
 *   returned text into a body. Higher quality and language-agnostic, at the
 *   cost of a model call.
 *
 * Either way the output is derived context (never a `MemoryEntry`, never a
 * `Proposal`) and is freely regenerable.
 */

export const SESSION_CONDENSER_VERSION = "pattern.v1";
export const LLM_CONDENSER_VERSION = "llm.v1";

/**
 * Recent messages are kept raw in the conversation window; everything older than
 * this tail is what `condenseSession` summarizes. Aligned with the chat
 * conversation-window recent-message limit so the raw tail and the summary
 * cover-range stay adjacent.
 */
export const DEFAULT_CONDENSE_KEEP_RECENT = 12;

/**
 * Re-condense only once at least this many new messages have aged past the last
 * summary's watermark. Bounds version churn to roughly one new summary per batch
 * instead of one per chat turn.
 */
export const DEFAULT_CONDENSE_BATCH = 8;

/** Defensive bound on the transcript handed to the LLM (chars). */
export const CONDENSE_PROMPT_MAX_CHARS = 12_000;

const MAX_HIGHLIGHTS = 6;
const HIGHLIGHT_MAX_CHARS = 160;
const MAX_KEYWORDS = 8;
const MIN_KEYWORD_LENGTH = 3;

// Small English stopword set; `pattern.v1` keyword extraction is intentionally
// ASCII-oriented and deterministic, not linguistically complete.
const STOPWORDS = new Set([
  "the", "and", "for", "are", "but", "not", "you", "your", "with", "this",
  "that", "have", "has", "had", "was", "were", "will", "would", "can", "could",
  "should", "from", "into", "about", "what", "when", "where", "which", "who",
  "how", "why", "all", "any", "did", "does", "doing", "done", "get", "got",
  "let", "lets", "its", "our", "out", "now", "use", "using", "than", "then",
  "they", "them", "their", "there", "here", "some", "such", "just", "like",
  "want", "need", "please", "thanks", "thank", "okay", "yes", "no",
]);

export interface CondenserMessage {
  id: string;
  role: string;
  content: string;
}

export interface SessionSummaryBody {
  summary_text: string;
  summary_json: Record<string, unknown>;
  source_first_message_id: string;
  source_last_message_id: string;
  source_message_count: number;
  token_estimate_before: number;
  token_estimate_after: number;
  condenser_version: string;
}

interface SourceStats {
  usable: CondenserMessage[];
  roleCounts: Record<string, number>;
  first: CondenserMessage;
  last: CondenserMessage;
  charsBefore: number;
}

/**
 * Deterministic `pattern.v1` body. `null` when the slice has no non-empty
 * content. The caller chooses which messages to pass (typically all messages
 * older than the recent raw tail).
 */
export function buildPatternSummary(
  messages: readonly CondenserMessage[],
): SessionSummaryBody | null {
  const stats = sourceStats(messages);
  if (!stats) return null;

  const keywords = topKeywords(stats.usable);
  const highlights = pickHighlights(stats.usable);
  const lines: string[] = [
    `Earlier conversation condensed (${stats.usable.length} message${
      stats.usable.length === 1 ? "" : "s"
    }${roleSummary(stats.roleCounts) ? `: ${roleSummary(stats.roleCounts)}` : ""}).`,
  ];
  if (keywords.length > 0) lines.push(`Topics: ${keywords.join(", ")}.`);
  if (highlights.length > 0) {
    lines.push("Highlights:");
    for (const highlight of highlights) {
      lines.push(`- ${highlight.role}: ${highlight.text}`);
    }
  }
  const summaryText = lines.join("\n");

  return summaryBody(stats, summaryText, SESSION_CONDENSER_VERSION, {
    top_keywords: keywords,
  });
}

/**
 * Condenser scenario profiles. The summary acts on chat-session messages, but a
 * session's content varies by who the user is talking to. `adaptive` (the
 * default) safely handles mixed personal + coding + project content; the others
 * specialize when the agent is dedicated to one scenario. Selected per agent via
 * `AgentVersion.context_policy_json.condenser.profile`; unknown / absent →
 * `adaptive`.
 */
export type CondenserProfile = "adaptive" | "general" | "coding" | "project";
export const DEFAULT_CONDENSER_PROFILE: CondenserProfile = "adaptive";

export interface CondenserPromptConfig {
  profile?: CondenserProfile | string | null;
}

export function resolveCondenserProfile(
  value: string | null | undefined,
): CondenserProfile {
  return value === "general" || value === "coding" || value === "project" || value === "adaptive"
    ? value
    : DEFAULT_CONDENSER_PROFILE;
}

/**
 * Wrap an LLM-produced summary string into an `llm.v1` body. `null` when the
 * slice has no content or the LLM returned empty text (so the caller falls back
 * to `pattern.v1`). `messages` is the full covered range — it sets the source
 * range / counts the summary claims to cover, independent of how many turns were
 * actually fed to the model.
 */
export function buildLlmSummary(
  messages: readonly CondenserMessage[],
  llmText: string,
): SessionSummaryBody | null {
  const stats = sourceStats(messages);
  if (!stats) return null;
  const text = llmText.trim();
  if (!text) return null;
  return summaryBody(stats, text, LLM_CONDENSER_VERSION, { generated_by: "llm" });
}

function sourceStats(messages: readonly CondenserMessage[]): SourceStats | null {
  const usable = messages.filter((message) => message.content.trim().length > 0);
  if (usable.length === 0) return null;
  const roleCounts: Record<string, number> = {};
  for (const message of usable) {
    roleCounts[message.role] = (roleCounts[message.role] ?? 0) + 1;
  }
  return {
    usable,
    roleCounts,
    first: usable[0]!,
    last: usable[usable.length - 1]!,
    charsBefore: usable.reduce((sum, m) => sum + m.content.trim().length, 0),
  };
}

function summaryBody(
  stats: SourceStats,
  summaryText: string,
  condenserVersion: string,
  extraJson: Record<string, unknown>,
): SessionSummaryBody {
  return {
    summary_text: summaryText,
    summary_json: {
      condenser_version: condenserVersion,
      role_counts: stats.roleCounts,
      source_range: {
        first_message_id: stats.first.id,
        last_message_id: stats.last.id,
        message_count: stats.usable.length,
      },
      ...extraJson,
    },
    source_first_message_id: stats.first.id,
    source_last_message_id: stats.last.id,
    source_message_count: stats.usable.length,
    token_estimate_before: estimateTokensFromChars(stats.charsBefore),
    token_estimate_after: estimateTokensFromChars(summaryText.length),
    condenser_version: condenserVersion,
  };
}

function roleSummary(roleCounts: Record<string, number>): string {
  return Object.keys(roleCounts)
    .sort()
    .map((role) => `${roleCounts[role]} from ${role}`)
    .join(", ");
}

function topKeywords(messages: readonly CondenserMessage[]): string[] {
  const counts = new Map<string, number>();
  for (const message of messages) {
    for (const token of tokenize(message.content)) {
      counts.set(token, (counts.get(token) ?? 0) + 1);
    }
  }
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1] || (a[0] < b[0] ? -1 : 1))
    .slice(0, MAX_KEYWORDS)
    .map(([word]) => word);
}

function tokenize(text: string): string[] {
  const matches = text.toLowerCase().match(/[a-z0-9]+/g);
  if (!matches) return [];
  return matches.filter(
    (word) => word.length >= MIN_KEYWORD_LENGTH && !STOPWORDS.has(word),
  );
}

function pickHighlights(
  messages: readonly CondenserMessage[],
): Array<{ role: string; text: string }> {
  // Prefer the user turns (they carry intent); fall back to the full transcript
  // when there are not enough user turns to fill the highlight budget. Either
  // way take the oldest first so the summary reads chronologically.
  const userTurns = messages.filter((message) => message.role === "user");
  const base = userTurns.length >= MAX_HIGHLIGHTS ? userTurns : messages;
  return base.slice(0, MAX_HIGHLIGHTS).map((message) => ({
    role: message.role,
    text: truncate(collapseWhitespace(message.content), HIGHLIGHT_MAX_CHARS),
  }));
}

function collapseWhitespace(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

function truncate(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;
  return `${text.slice(0, Math.max(0, maxChars - 1)).trimEnd()}…`;
}

function estimateTokensFromChars(chars: number): number {
  return chars > 0 ? Math.ceil(chars / 4) : 0;
}
